### PR TITLE
BENCH-143: Update Add Category to Product

### PR DESCRIPTION
**Pull request by bethcryer**
You can now update the collection of products in each category through the Category POST and PUT endpoints.
You can no longer do so through the Product POST and PUT endpoints.

Tests have been updated, and I moved some over from ProductServiceTests to CategoryServiceTests since they're now applicable there instead.

### CODE REVIEW

File: src/Answer.King.Api/Validators/CategoryValidator.cs [hunk 1]
```diff
@@ -12,5 +12,10 @@ public CategoryValidator()
 
         this.RuleFor(c => c.Description)
             .NotNullOrWhiteSpace();
+
+        this.RuleFor(p => p.Products)
```
Comment (milkshakeuk) [inline src/Answer.King.Api/Validators/CategoryValidator.cs:16]: We probably want a `.NotEmpty()` before the `.ForEach`

File: src/Answer.King.Domain/Inventory/Category.cs [hunk 2]
```diff
@@ -5,7 +5,7 @@ namespace Answer.King.Domain.Inventory;
 // Todo: look at custom deserialisation: https://stackoverflow.com/questions/42336751/custom-deserialization
 public class Category : IAggregateRoot
 {
-    public Category(string name, string description)
+    public Category(string name, string description, IList<ProductId> products)
     {
         Guard.AgainstNullOrEmptyArgument(nameof(name), name);
         Guard.AgainstNullOrEmptyArgument(nameof(description), description);
```
Comment (milkshakeuk) [inline src/Answer.King.Domain/Inventory/Category.cs:11]: We need to guard against null for products here.

File: src/Answer.King.Domain/Repositories/Models/Product.cs [hunk 3]
```diff
@@ -27,7 +26,6 @@ public Product(string name, string description, double price, IList<CategoryId>
         Guard.AgainstNullOrEmptyArgument(nameof(name), name);
         Guard.AgainstNullOrEmptyArgument(nameof(description), description);
         Guard.AgainstNegativeValue(nameof(price), price);
-        Guard.AgainstNullOrEmptyArgument(nameof(categories), categories);
```
Comment (milkshakeuk) [inline src/Answer.King.Domain/Repositories/Models/Product.cs:30]: do we at least need a guard for null for categories? I think we do.

File: src/Answer.King.Api/Validators/CategoryValidator.cs (hunk 1 above)
Comment (bethcryer) [reply src/Answer.King.Api/Validators/CategoryValidator.cs:16]: Do we always want new Categories to contain at least one product?

File: src/Answer.King.Domain/Inventory/Category.cs (hunk 2 above)
Comment (bethcryer) [reply src/Answer.King.Domain/Inventory/Category.cs:11]: done

File: src/Answer.King.Domain/Repositories/Models/Product.cs (hunk 3 above)
Comment (bethcryer) [reply src/Answer.King.Domain/Repositories/Models/Product.cs:30]: yup added it back in

File: src/Answer.King.Api/Validators/CategoryValidator.cs (hunk 1 above)
Comment (milkshakeuk) [reply src/Answer.King.Api/Validators/CategoryValidator.cs:16]: good point probably not.

File: src/Answer.King.Api/Services/CategoryService.cs [hunk 4]
```diff
@@ -2,18 +2,27 @@
...
     {
         this.Categories = categories;
+        this.Products = products;
     }
 
     private ICategoryRepository Categories { get; }
 
+    private IProductRepository Products { get; }
+
+
```
Comment (milkshakeuk) [inline src/Answer.King.Api/Services/CategoryService.cs]: extra carriage return needs removing.

File: src/Answer.King.Domain/Orders/Models/Product.cs [hunk 5]
```diff
@@ -10,7 +10,6 @@ public Product(long id, string name, string description, double price, IList<Cat
         Guard.AgainstNullOrEmptyArgument(nameof(name), name);
         Guard.AgainstNullOrEmptyArgument(nameof(description), description);
         Guard.AgainstNegativeValue(nameof(price), price);
-        Guard.AgainstNullOrEmptyArgument(nameof(categories), categories);
```
Comment (milkshakeuk) [inline src/Answer.King.Domain/Orders/Models/Product.cs:13]: I think we need to guard against null categories here.

File: tests/Answer.King.Api.IntegrationTests/Common/Models/Product.cs [hunk 6]
```diff
@@ -10,7 +10,6 @@ public Product(long id, string name, string description, double price, IList<lon
         Guard.AgainstNullOrEmptyArgument(nameof(name), name);
         Guard.AgainstNullOrEmptyArgument(nameof(description), description);
         Guard.AgainstNegativeValue(nameof(price), price);
-        Guard.AgainstNullOrEmptyArgument(nameof(categories), categories);
```
Comment (milkshakeuk) [inline tests/Answer.King.Api.IntegrationTests/Common/Models/Product.cs:13]: do we even need any guards in the integration tests models?

**Review by milkshakeuk — CHANGES_REQUESTED** [review]
couple of comments.

File: tests/Answer.King.Api.IntegrationTests/Common/Models/Product.cs (hunk 6 above)
Comment (bethcryer) [reply tests/Answer.King.Api.IntegrationTests/Common/Models/Product.cs:13]: good shout, have now removed those guards and also the setters that weren't being used

File: src/Answer.King.Domain/Orders/Models/Product.cs (hunk 5 above)
Comment (bethcryer) [reply src/Answer.King.Domain/Orders/Models/Product.cs:13]: done

File: src/Answer.King.Api/Services/CategoryService.cs (hunk 4 above)
Comment (bethcryer) [reply src/Answer.King.Api/Services/CategoryService.cs]: done

**Review by milkshakeuk — APPROVED** [review]

**Review by kirbyAlasdairAnswer — APPROVED** [review]

**Pull request merged**